### PR TITLE
Permit access to select environment where net-cap-bind support active

### PR DESCRIPTION
**Pull request by planetf1**
Here's a first version of my PR for issue #21770.

Hopefully the overall approach makes sense. Very much welcome feedback!

## User Scenario

A user executing an application on linux that utilizes open ssl libraries cannot
make use of quantum-safe encryption algorithms, using **OPENSSL_MODULES** and **OPENSSL_CONF**, if that application needs the ability
to bind to a port *below 1024* by enabling **CAP_NET_BIND_SERVICE**.

Setting this privilege is common in server environments to run on well known ports.

## Technical Problem Summary

When using GLIBC \> 2.17 (linux?), OpenSSL uses the `secure_getenv` call in glibc to retrieve all environment variables. If the process is setuid/setgid, OR has additional capabilities set, this call will return NULL. 

This is intentional, since we do not want a non-priviliged user to inject malicious code, or otherwise compromise the configuration of the application.

## Proposed change of behavior

The exact test as to whether to return NULL is slightly modified, in that it is only applied IF, and ONLY IF, the ONLY additional capability granted is **CAP_NET_BIND_SERVICE**.

## Risks

The risks to evaluate for this solution include:
- An error in the design such that the intended test allows for malicious configuration. 
- An error in the implementation such that the code does not implement the intended design. 
- Poor documentation, such that the feature is unknown, or unclear to use
- Library/kernel changes affecting the implementation behaviour in future

## Proposed Implementation in this PR

The intent is to keep checks as safe as possible, and make minimal changes to existing code.

The changes are protected by compile time flags, and the restriction is only lifted for SELECT environment variables, with ONLY the specified capability.

- Add new compile-time define **OPENSSL_NETCAP_ALLOW_ENV**, so that the modified code is only compiled if specified at compile time ie using `CFLAGS=-DOPENSSL_NETCAP_ALLOW_ENV` (all)
- If set, and we are on linux (**__linux__** defined) uses new function `restricted_getenv()` to retrieve specified environment variable.  Question - is the preprocessor check here appropriate? (*getenv.c*)
- `restricted_getenv``  uses a modified **OpenSSL_issetugid** (below) to check we aren't in a secure mode. This is a change in behaviour from previously, where this function was only used if glibc was too old to support `secure_getenv` (*getenv.c*)
- **OpenSSL_issetugid** is modified slightly so that it queries the processes's capabilities if we are in a secure mode. If the only capability set is for bind, we now return false (0) . This follows the intent of the function to indicate we need to act very securely in general ie for setuid etc., but allows for this proposed loosening of the check. (*uid.c*)
- `restricted_getenv` then checks if the requested environment variable is a member of a configured (hardcoded) list (*getenv.c*)
- If so, it is returned directly. If not it reverts to using `secure_getenv`.

## Build

For a 'local' development build to support testing, making use of a pause check at startup (see below).

Install into a local directory (not overwriting system version)
```shell
./Configure --prefix=/home/jonesn/opt/openssl --openssldir=/home/jonesn/usr/local/ssl
```

Build with appropriate flags:
```shell
make CFLAGS=-DOPENSSL_NETCAP_ALLOW_ENV -g -Wl,-rpath,/home/jonesn/src/openssl
```
## Alternative implementations considered

There are many ways of setting up the preprocessor checks, and exact code sequence  -- and the flow could likely be made simpler, but it does result in more changes, so more risk.

There are also some utility functions in libcap. However it is hard to check for only a select, permitted capability with the limited API. Additionally it would introduce an additional development and runtime dependency, albeit a fairly standard one.

### Making this feature available by default

OpenSSL libraries are used by many applications. We need to make it easier for users of these applications to enable quantum-safe cryptography schemes by deployment/app configuration, and not requiring a rebuild of openssl libraries.

To do this, I would like to suggest that if this PR is acceptable to the reviewers, that either within this PR, or in a later PR we can either
 - remove the build flag entirely
 - default the value to being enabled

### Runtime configuration of this feature

* Do we need to allow this feature to be disabled at runtime, in the configuration
* Can we make the list of environment variables permitted configurable? 

Both have one notable issue - where to store the config, especially if we already allow OPENSSL_CONF to be passed

* Is it safe to permit all?

## Testing

Recommendations on the best way to test welcomed. 

So far testing has been done by using the 'openssl' command, and validating if environment variables are passed by inspection of command output, or by using a debugger (remotely attached so as not to affect permissions).

For example
```shell
setcap cap_net_bind_service=+ep  apps/openssl
export OPENSSL_CONF=/home/jonesn/etc/openssl.cnf 
./apps/openssl engine -c qatengine
```

Capability was added to the `apps/openssl` executable where needed for the test using `setcap cap_net_bind_service=+ep  apps/openssl`. Other capabilities can also be set to verify these environment variables are no longer allowed to pass.

It was not possible to test (in Visual Studio Code) using gdb directly, as if run under debug no extra capabilities set set. Neither can the process be run as root, since then it has too many permissions. Instead the `apps/openssl` app was run, and then attached to remotely, using gdb running as root.

To do this the openssl command was paused at startup using this at the top of *apps/openssl.c*:
```c
#ifdef OPENSSL_TEST_PAUSE_FOR_ATTACH
#include <signal.h>
#endif
``````
then later, at the beginning of main:
```c
#if defined(OPENSSL_TEST_PAUSE_FOR_ATTACH)
    if(getenv("OPENSSL_TEST_PAUSE_FOR_ATTACH"))
    {
        raise(SIGSTOP);
    }
#endif
```
This can be used by defining appropriate values at build/runtime.

If this is useful I am happy to submit further changes, but it was very much just
a mechanism to support my testing given the complexity of capabilities. It does not form part of the current PR. Similarly I can add more docs on visual studio code build/test if it helps.

I'll continue testing, to verify that the user scenario works, i.e. using the  liboqs provider & libraries:
- https://github.com/open-quantum-safe/liboqs
- https://github.com/open-quantum-safe/oqs-provider

## Documentation

There should be some end-user documentation of this feature. NOTES-UNIX.md may be one location?
Should information be added elsewhere?

### CODE REVIEW

**Comment by planetf1** [comment]
Here's a first version of my PR for issue #21770.

Hopefully the overall approach makes sense. Very much welcome feedback!

## User Scenario

A user executing an application on linux that utilizes open ssl libraries cannot
make use of quantum-safe encryption algorithms, using **OPENSSL_MODULES** and **OPENSSL_CONF**, if that application needs the ability
to bind to a port *below 1024* by enabling **CAP_NET_BIND_SERVICE**.

Setting this privilege is common in server environments to run on well known ports.

## Technical Problem Summary

When using GLIBC \> 2.17 (linux?), OpenSSL uses the `secure_getenv` call in glibc to retrieve all environment variables. If the process is setuid/setgid, OR has additional capabilities set, this call will return NULL. 

This is intentional, since we do not want a non-priviliged user to inject malicious code, or otherwise compromise the configuration of the application.

## Proposed change of behavior

The exact test as to whether to return NULL is slightly modified, in that it is only applied IF, and ONLY IF, the ONLY additional capability granted is **CAP_NET_BIND_SERVICE**.

## Risks

The risks to evaluate for this solution include:
- An error in the design such that the intended test allows for malicious configuration. 
- An error in the implementation such that the code does not implement the intended design. 
- Poor documentation, such that the feature is unknown, or unclear to use
- Library/kernel changes affecting the implementation behaviour in future

## Proposed Implementation in this PR

The intent is to keep checks as safe as possible, and make minimal changes to existing code.

The changes are protected by compile time flags, and the restriction is only lifted for SELECT environment variables, with ONLY the specified capability.

- Add new compile-time define **OPENSSL_NETCAP_ALLOW_ENV**, so that the modified code is only compiled if specified at compile time ie using `CFLAGS=-DOPENSSL_NETCAP_ALLOW_ENV` (all)
- If set, and we are on linux (**__linux__** defined) uses new function `restricted_getenv()` to retrieve specified environment variable.  Question - is the preprocessor check here appropriate? (*getenv.c*)
- `restricted_getenv``  uses a modified **OpenSSL_issetugid** (below) to check we aren't in a secure mode. This is a change in behaviour from previously, where this function was only used if glibc was too old to support `secure_getenv` (*getenv.c*)
- **OpenSSL_issetugid** is modified slightly so that it queries the processes's capabilities if we are in a secure mode. If the only capability set is for bind, we now return false (0) . This follows the intent of the function to indicate we need to act very securely in general ie for setuid etc., but allows for this proposed loosening of the check. (*uid.c*)
- `restricted_getenv` then checks if the requested environment variable is a member of a configured (hardcoded) list (*getenv.c*)
- If so, it is returned directly. If not it reverts to using `secure_getenv`.

## Build

For a 'local' development build to support testing, making use of a pause check at startup (see below).

Install into a local directory (not overwriting system version)
```shell
./Configure --prefix=/home/jonesn/opt/openssl --openssldir=/home/jonesn/usr/local/ssl
```

Build with appropriate flags:
```shell
make CFLAGS=-DOPENSSL_NETCAP_ALLOW_ENV -g -Wl,-rpath,/home/jonesn/src/openssl
```
## Alternative implementations considered

There are many ways of setting up the preprocessor checks, and exact code sequence  -- and the flow could likely be made simpler, but it does result in more changes, so more risk.

There are also some utility functions in libcap. However it is hard to check for only a select, permitted capability with the limited API. Additionally it would introduce an additional development and runtime dependency, albeit a fairly standard one.

## Future enhancements

### Making this feature available by default

OpenSSL libraries are used by many applications. We need to make it easier for users of these applications to enable quantum-safe cryptography schemes by deployment/app configuration, and not requiring a rebuild of openssl libraries.

To do this, I would like to suggest that if this PR is acceptable to the reviewers, that either within this PR, or in a later PR we can either
 - remove the build flag entirely
 - default the value to being enabled

### Runtime configuration of this feature

* Do we need to allow this feature to be disabled at runtime, in the configuration
* Can we make the list of environment variables permitted configurable? 

Both have one notable issue - where to store the config, especially if we already allow OPENSSL_CONF to be passed

* Is it safe to permit all?

## Testing

Recommendations on the best way to test welcomed. 

So far testing has been done by using the 'openssl' command, and validating if environment variables are passed by inspection of command output, or by using a debugger (remotely attached so as not to affect permissions).

For example
```shell
setcap cap_net_bind_service=+ep  apps/openssl
export OPENSSL_CONF=/home/jonesn/etc/openssl.cnf 
./apps/openssl engine -c qatengine
```

Capability was added to the `apps/openssl` executable where needed for the test using `setcap cap_net_bind_service=+ep  apps/openssl`. Other capabilities can also be set to verify these environment variables are no longer allowed to pass.

It was not possible to test (in Visual Studio Code) using gdb directly, as if run under debug no extra capabilities set set. Neither can the process be run as root, since then it has too many permissions. Instead the `apps/openssl` app was run, and then attached to remotely, using gdb running as root.

To do this the openssl command was paused at startup using this at the top of *apps/openssl.c*:
```c
#ifdef OPENSSL_TEST_PAUSE_FOR_ATTACH
#include <signal.h>
#endif
``````
then later, at the beginning of main:
```c
#if defined(OPENSSL_TEST_PAUSE_FOR_ATTACH)
    if(getenv("OPENSSL_TEST_PAUSE_FOR_ATTACH"))
    {
        raise(SIGSTOP);
    }
#endif
```
This can be used by defining appropriate values at build/runtime.

If this is useful I am happy to submit further changes, but it was very much just
a mechanism to support my testing given the complexity of capabilities. It does not form part of the current PR. Similarly I can add more docs on visual studio code build/test if it helps.

I'll continue testing, to verify that the user scenario works, i.e. using the  liboqs provider & libraries:
- https://github.com/open-quantum-safe/liboqs
- https://github.com/open-quantum-safe/oqs-provider

## Documentation

There should be some end-user documentation of this feature. NOTES-UNIX.md may be one location?
Should information be added elsewhere?

File: crypto/getenv.c [hunk 1]
```diff
@@ -15,6 +15,50 @@
...
+  return 0;
+}
+
+char *restricted_getenv(const char *name)
+{
+    /* Environment variables to permit in secure mode (subject to conditions). Extend as needed */
+    const char * permitted_env [ ] = { "OPENSSL_MODULES", "OPENSSL_CONF" };
+    if(name!=NULL && (!OPENSSL_issetugid()))
+    {
+      if (str_in_list(name,permitted_env,sizeof(permitted_env)/sizeof(char *)))
```
Comment (tom-cosgrove-arm) [inline crypto/getenv.c:52]: Rather than hard-code the type (`char *`), please use something like

```
#define OSSL_NELEM(x) (sizeof(x)/sizeof((x)[0]))
```

File: crypto/getenv.c [hunk 2]
```diff
@@ -15,6 +15,50 @@
...
+  if ( (candidate!=NULL) && (list!=NULL))
+  {
+    int len = n;
+    if (len>=1)
+    {
+        int i;
+        for ( i=0; i<len; i++)
+        {
+            if (strcmp(candidate,list[i])==0)
+              // found
```
Comment (tom-cosgrove-arm) [inline crypto/getenv.c:37]: Not really sure this comment is helpful

File: crypto/getenv.c [hunk 3]
```diff
@@ -15,6 +15,50 @@
...
+ * Additional checks for securely allowing access to selected environment variables
+ */
+#if defined (OPENSSL_NETCAP_ALLOW_ENV) && defined(__linux__)
+
+/*
+ * Utility function to check if a string is in an array of strings. returns true if found
+ */
+int str_in_list(const char *candidate, const char *list[], int n)
+{
+  if ( (candidate!=NULL) && (list!=NULL))
```
Comment (tom-cosgrove-arm) [inline crypto/getenv.c:28]: Rather than having "arrow code", why not test inputs and do an early return

```
if (candidate == NULL || list == NULL || len < 1)
    return 0;
```

File: crypto/getenv.c [hunk 4]
```diff
@@ -91,7 +135,18 @@ char *ossl_safe_getenv(const char *name)
...
 #  define SECURE_GETENV
-    return secure_getenv(name);
+#   /* 
+     * Non-default option to permit selected additional variables to be read where 
+     * only the netcap capability is added, rather than follow the standard
+     * priviliged default of disallowing access. (libc > 2.17, linux only)
+     */
+#   if defined(OPENSSL_NETCAP_ALLOW_ENV) && defined(__linux__)
+     /* Use alternate implementation which allows some exceptions if in a secure environment */
+       return(restricted_getenv(name));
```
Comment (tom-cosgrove-arm) [inline crypto/getenv.c:145]: This line doesn't align with the comment above.

In general, there are lots of whitespace issues, including
- missing space after `,`
- missing space either side of binary operators
- opening braces should be on the same line as `if`, `for`, etc
- unnecessary whitespace before `[]` and between empty `[]`
- unnecessary whitespace after `*` in `char * permitted_env`

File: crypto/getenv.c (hunk 1 above)
Comment (paulidale) [reply crypto/getenv.c:52]: The `OSSL_NELEM` macro is defined in `internal/nelem.h`.

File: crypto/getenv.c [hunk 5]
```diff
@@ -91,7 +135,18 @@ char *ossl_safe_getenv(const char *name)
 #if defined(__GLIBC__) && defined(__GLIBC_PREREQ)
 # if __GLIBC_PREREQ(2, 17)
 #  define SECURE_GETENV
-    return secure_getenv(name);
+#   /*
```
Comment (nhorman) [inline crypto/getenv.c:138]: is that # supposed to be there?  I've not tried but it seems like that won't compile

File: crypto/getenv.c [hunk 6]
```diff
@@ -15,6 +15,50 @@
...
+    }
+  }
+  // drop-through case -- not found (or invalid parameters)
+  return 0;
+}
+
+char *restricted_getenv(const char *name)
+{
+    /* Environment variables to permit in secure mode (subject to conditions). Extend as needed */
+    const char * permitted_env [ ] = { "OPENSSL_MODULES", "OPENSSL_CONF" };
```
Comment (nhorman) [inline crypto/getenv.c:49]: This feels a bit like we're hardcoding policy to me (which may or may not be ok).  Currently, ossl_safe_getenv is called to fetch the following environment variables:

OPENSSL_CONF_INCLUDE
CTLOG_FILE_EVP
OPENSSL_CONF
OPENSSL_ENGINES
no_proxy
OPENSSL_NO_PROXY
OPENSSL_HTTP_PROXY
OPENSSL_HTTPS_PROXY
OPENSSL_MODULES
RANDFILE
HOME
X509_CERT_DIR_EVP

Seeing this makes me ask "which of those variables do we want to allow to be fetched when in a secure environment, which do we want to hide, and under what conditions?"

The more I look at it, the more I think the only reasonable answer is "we don't know" or more specifically "we can't know".  I can see how this works in your situation because you specifically need those environment variables, but I could just as easily see another user specifically wanting to restrict access to them when in a secure environment, or they may need a superset of them, some of which your use case may consider to be a security risk.  As such, it seems wrong to hardcode them.

A better solution might be to create an OPENSSL_ALLOW_ENV variable that can be read by ossl_safe_getenv which holds a list of env vars that should be fetched using getenv rather than secure_getenv.  That would allow users to set a selected list of trusted variables at run time to be passed through from outside the secure environment.

Of course, such an implementation would require that OPENSSL_ALLOW_ENV either be retrieved via getenv (which negates the security introduced by secure_getenv), or require that it be set within the secure environment, which puts you back to needing a wrapper script.

File: crypto/getenv.c (hunk 2 above)
Comment (planetf1) [reply crypto/getenv.c:37]: sure. will remove.

File: crypto/getenv.c (hunk 5 above)
Comment (planetf1) [reply crypto/getenv.c:138]: No. However it compiles for me, and with those definitions active (thus using my new code). Odd. However I've removed and will retest. thanks!

File: crypto/getenv.c (hunk 6 above)
Comment (planetf1) [reply crypto/getenv.c:49]: This is a critical design question. 

The intent of this draft was to demonstrate a mechanism, but also point out the exact list needs discussion, and could be configuration dependent.

It cannot be an environment variable (at least with the current design) since that effectively would allow all values. We have a specific issue with the configuration needed for quantum-safe encryption, therefore I decided to place a hardcoded set initially. I agree that ultimately a hard-coded set is not desirable.

Note that these extra variables should not be read in all secure environments, only if the ONLY capability set is to allow ports <1024 to be allocated. In other cases they should be hidden.

The list could be defined in openssl.cnf. However if the default did not permit these values we'd be back to needing deployment-time configuration.?

Would we be able to get consensus on a default set?

Comment (nhorman) [reply crypto/getenv.c:49]: Thats the thing I'm strugging with here, I don't think we can agree on a default set, because eventually, someone will say "Can we add the variable X to the list?".  And perhaps we can, or decide we can't, but its going to be a game of whack-a-mole one way or ther other, explaining to subsequent users why we think their variable can or can't be on the list.  The only way I see to avoid that, is to let users decide whats safe in their environment, for which we either need an allowlist environment variable (with the recursive inside/outside the secure boundary issues that brings with it) or a command line --allow-env option on any application that wants to make use of this feature.

The capabilies set is the same problem in my mind. restricting this feature to only work when CAP_NET_BIND_SERVICE is the only capability set makes sense for this use case, but there are lots of capabilities that I could envision users wanting to set that would also necessitate the use of this mechanism.  It makes more sense in my mind to use a mechanism like this always (i.e. assume we are in a secure environment, and if secure_getenv fails, back off to using getenv, but only for those variables which a user has informed us at run time are considered safe to fetch from outside the secure environment

File: crypto/getenv.c (hunk 1 above)
Comment (planetf1) [reply crypto/getenv.c]: Thanks for the tip - have refactored as per suggestion
(I've left explicit conditional checks in at this stage)

File: crypto/getenv.c (hunk 5 above)
Comment (planetf1) [reply crypto/getenv.c]: fixed

File: crypto/getenv.c (hunk 4 above)
Comment (planetf1) [reply crypto/getenv.c:143]: I've made the changes I've noticed. Will keep this conversation open and review https://www.openssl.org/policies/technical/coding-style.html - thanks for the reminder.

File: crypto/getenv.c [hunk 7]
```diff
@@ -15,6 +15,48 @@
...
+    if (strcmp(candidate, list[i]) == 0)
+      return 1;
+  }
+  return 0;
+}
+
+char *restricted_getenv(const char *name)
+{
+    /* Environment variables to permit in secure mode (subject to conditions). Extend as needed */
+    const char *permitted_env[] = { "OPENSSL_MODULES", "OPENSSL_CONF" };
```
Comment (juergenchrist) [inline crypto/getenv.c:47]: This seems like a very bad idea.  Those are the two variables that allow you to specify what code to load by default (via provider loading and activation) and give you the means to redirect this to a place where you can put your code.  So this makes OpenSSL kind of insecure in this case.

File: crypto/getenv.c (hunk 6 above)
Comment (paulidale) [reply crypto/getenv.c]: Seems the options available to specify which are exempt are:

* API call
* Command line argument
* Configuration
* Environment variable (although the consensus seems to be that this is a bad idea)
* Hard coded list (too difficult to define globally)

File: crypto/getenv.c (hunk 7 above)
Comment (planetf1) [reply crypto/getenv.c:47]: I understand the concern -- putting aside the questions around configurability,  the core proposal is specific to net-cap-bind - indeed it should only be a change where this is the ONE and ONLY capability set. Any more, and the test fails, and we use secure_getenv(). (I added a PR after the initial issue discussion so as to be explicit about what was proposed)

As you say, it does allow for injection of the providers (by spec of directory, and configuration) - which is where the risk of malicious code lies.

This capability gives the ability for a process to bind to a port < 1024, nothing else. So the risk is of injected code becoming active on those ports. This is the core question - is that reasonable?

The 'node' project made a similar change in https://github.com/nodejs/node/pull/37727 - of course that's a different project, with a different role

Comment (t8m) [reply crypto/getenv.c:47]: There is no point in allowing just these two variables and routing the rest through secure_getenv(). Just use this for any ossl_safe_getenv() calls.

And of course all this is behing  OPENSSL_NETCAP_ALLOW_ENV ifdef so this is not built by default. So IMO ok.

Comment (planetf1) [reply crypto/getenv.c:47]: @juergenchrist I am certainly concerned about avoiding opening up security holes.

The core proposal was focussed specifically on net-cap-bind capability, which allows a process to bind to a port <1024. If any other capabilities are found, this new logic should not apply, regardless of net-cap-bind being set,and environment variables blocked .

The potential exposure with this capability would be to allow malicious or unexpected code to become active on a port <1024 where previously they could not, in the case where an admin had granted the capability, but does not control the values of the environment ie provider, libraries etc. Today the combination of binding to ports < 1024 & having control of the configuration through variables is not possible by a non-admin user.

The 'node' project opted to permit the loosening of the check, albeit that is one specific application. I created the PR to be explicit about the proposal, to see if the community would find this restricted loosening  appropriate. The discussion above also has touched on config, but it's this specific case where I see this level of appropriateness.

The overall intent is to enable the use of new quantum safe encryption algorithms (ie as provided by the liboqs project and liboqs openssl provider) to be easier. This can be done via admin configuration of openssl.cnf/placing libraries in appropriate location,  via the app utilizing openssl, or (sometimes fiddly) wrapping techniques. We're trying to see if we can avoid those steps.

Long term perhaps these new algorithms will make it into core openssl, at which point maybe issue won't apply, but even then crypto-agility is important, so it will then just be a new set of quantum-safe algorithms.... !

File: crypto/getenv.c (hunk 6 above)
Comment (nhorman) [reply crypto/getenv.c]: To be complete, there is another option: do nothing.  Its still entirely possible to wrap an application in a capabilities adjusted shell wrapper that explicitly sets any needed environment variables and execs the needed openssl linked application. Doing so gives the user the most control over what environment variables are allowed in and under what conditions they feel comfortable with

File: crypto/uid.c [hunk 8]
```diff
@@ -44,10 +49,50 @@ int OPENSSL_issetugid(void)
...
+        return cap_data[0].permitted == (CAP_TO_MASK(capability));
+    }
+    // Probably also need to check [1] - some capabilities are >32
+    // to check for ONLY net_bind we need to ensure these are 0 also
+    
+    return cap_data[1].permitted == (CAP_TO_MASK(capability));
+}
+#endif
+
 int OPENSSL_issetugid(void)
```
Comment (mikebrow) [inline crypto/uid.c:87]: nit..overloading the bool check this way hides the impl somewhat for OPENSSL_issetugid when both OSSL_IMPLEMENT_GETAUXVAL and OPENSSL_NETCAP_ALLOW_ENV are defined.. suggest keeping use of HasOnlyCapability(CAP_NET_BIND_SERVICE) separate from the u/gid check or maybe rename OPENSSL_allowlimtedenvcopy(void) ..

Comment (planetf1) [reply crypto/uid.c:87]: Thanks, that's a good suggestion.

File: crypto/getenv.c (hunk 6 above)
Comment (planetf1) [reply crypto/getenv.c]: Why did we think this approach could be useful?

There are many services listening on standard (<1024) ports which make use of OpenSSL. This might include apps like postgres, nodejs, hginx, haproxy, openssh, a wide range of https applications. Some of these apps themselves are used in front of others (on >1024) as proxies/routers

Each of these will differ in terms of the process model - which would affect how wrappers could be introduced.

Each are likely working on specific plans to use new encryption algorithms, but this will take time.

Having a common approach across all these tools -- for this specific case (there is no issue > 1024) -- would allow quicker adoption of quantum-safe encryption algorithms, with likely no app change, by appropriately configuring those algorithms & making the executable modules available through the environment variables listed - as can already be done for apps on ports >1024 thanks to the provider support in 3.x, and implementations such as liboqs. 
Since my understanding is that the net bind capability is a very specific capability, that this would be ok. 

Wrappers, authbind, application-specific coding, system-wide configurations remain other options on a by application/deployment basis

Comment (nhorman) [reply crypto/getenv.c]: I agree this is useful, I have no  argument there.

My concern is for whom else this might be useful (or more specifically useful if we just expand the capabilities for which it can be used).

To cite your example of applications which set CAP_NET_BIND_SERVICE, looking at nginx, it also has code that sets CAP_NET_RAW when NGX_HAVE_CAPABILITIES is defined in that build.  In such a build, this change becomes "non-useful", in that OPENSSL_* variables become hidden again, when they otherwise might want them.

So, the question: Do we add CAP_NET_RAW to the filter list, effectively saying only those two are allowed?  Or do we say no, only CAP_NET_BIND_SERVICE, because we think thats safe?

My argument is that we shouldn't be the gatekeepers of what we think is safe in this case.  If we're going to use secure_getenv, we need to use it, as it protects us from malicious intrusion via injected environment variables.

If a user of openssl thinks something is safe, thats their business, but they need work with the environment that the os is giving us.  I understand that will be different for different deployments, but thats the only way to place this policy in the hands of the users making that decision.

If we want to help them manage that, I think thats fine and good, but we shouldn't code that policy into the library itself.  I think the best we can do is either:
a) guide users in how to setup the secure environment properly (via a wrapper script or some such)
or
b) give them some tools to implement such a policy on their own (i.e. adding a command line switch to apps we distribute to dynamically create an alowlist of variables they consider safe to read from outside the secure environment at their own risk)

**Comment by beldmit** [comment]
Close/reopen to trigger CLA check

File: crypto/getenv.c (hunk 6 above)
Comment (planetf1) [reply crypto/getenv.c]: @nhorman Good questions, and that's an interesting observation on NGINX. I agree configuration needs to be in the control of the user (the person building the code for now).

I'm just delaying further updates briefly which I go back to check out use cases, particularly in the context of wrappers (and/or authbind) before looking at the remaining comments, and in particular a proposal around configurability

**Comment by t8m** [comment]
**OTC: We cannot accept this PR. We do not want to hardcode exemption for specific environment variables depending on Linux-specific syscalls or Linux capability APIs. We would consider a mechanism which allows an application to specify a list of environment variable names which should not be routed via secure getenv.**

**Comment by planetf1** [comment]
> **OTC: We cannot accept this PR. We do not want to hardcode exemption for specific environment variables depending on Linux-specific syscalls or Linux capability APIs. We would consider a mechanism which allows an application to specify a list of environment variable names which should not be routed via secure getenv.**

Thanks for the clarification and feedback. I will work on a new/updated PR which allows for the environment variables to be configured at build time, and continue the discussion in #21770
